Initial load of ASP.NET MVC and Angular post. (#2603)
okta/okta.github.io: _source/_posts/2018-12-20-build-basic-web-app-with-mvc-angular.md

After finishing, you should have a clean project and perfect starting ground.

## Create an Angular Project
Using Angular CLI is the best way to get started with Angular applications. With Angular CLI tools, you can easily create new projects, update existing projects, add different components to your codebase, and do all of that by entering a few commands in your terminal. You can also run the application in the development environment and publish it to production environments. It saves you the trouble of setting up configuration and Webpack. It can also do all the optimizing, compressing and minifying for you. It's highly recommended to use Angular CLI to prepare your applications for production environments.

You can install the Angular CLI by entering the following in your favorite terminal:

```sh
npm install -g @angular/cli@7.1.0
```

After that, navigate to the folder of your MVC project and open your terminal there and enter the following:

```sh
ng new Angular --skip-tests --style=scss
```

This will create a new Angular project under Angular folder within your MVC project. The `--skip-tests` argument will make sure that CLI skips creating spec files.

> You should include the Angular folder in your MVC project. To do that, click on the `Show All Files` button in your Solution Explorer. You can now navigate to `Angular ` folder and include all folders and files except for the `node_modules`. After you do that you should have a folder structure like on the following picture:

Reviewer: ```suggestion
You should include the Angular folder in your MVC project. To do that, click on the **Show All Files** button in your Solution Explorer. You can now navigate to **Angular** folder and include all folders and files except for `node_modules`. After you do that you should have a folder structure like on the following picture:
```

## Set up a Database for ASP.NET
You will use Entity Framework (EF) 6 as your Object-Relational Mapper (ORM), a proven solution from Microsoft. EF 6 is pretty easy to set up and get going, and you don't need to know any SQL to have it working for your applications. Think of it as an adapter or bridge to your database. All of your database queries will go through Entity Framework. 

The central point is your `DbContext` class, and that's where you define your connection string and the tables for your database. EF 6 uses the `DbSet` collection type to represent a table in a database. All you need to do is create a connection string, create a new class that inherits from `DbContext`, make the data models and create appropriate properties within `DbContext` for those models. Those properties will represent tables in the database. From that point, it's pretty easy to get and update existing data by using Language Integrated Queries (LINQ).

You should first install the EF 6, by running the following code in the Package Manager Console:

```sh
Install-Package EntityFramework -Version 6.2.0
```

### Add the Connection String

> You can get started by adding a connection string to your `Web.config` file. Make sure to add it inside of the `<configuration>` section, and after `<configSections></configSections>`

Reviewer: ```suggestion
You can get started by adding a connection string to your `Web.config` file. Make sure to add it inside of the `<configuration>` section, and after `<configSections></configSections>`:
```

```xml
<connectionStrings>
  <add name="OktaConnectionString"
    connectionString="Data Source=(LocalDb)\MSSQLLocalDB;Initial Catalog=JoggingTracker;Integrated Security=SSPI;AttachDBFilename=|DataDirectory|\WeightTracker.mdf"
    providerName="System.Data.SqlClient"
  />
</connectionStrings>
```

### Create Your DbContext and Data Models in Angular
First, create a model for a recording a jog. Create a file in the `Models` directory called `JoggingRecord.cs`:

```cs
using System;

namespace OktaAspNetMvcAngular.Models
{
  public class JoggingRecord
  {
    public int Id { get; set; }

    public float Distance { get; set; }

    public string Description { get; set; }

    public DateTime CreatedAt { get; set; }
  }
}
```

Nothing special here, a standard `Id` property that represents a unique identifier for every row within a table and Entity Framework will use a naming convention and automatically set it as a primary key for this table. You definitely need the `Distance` property to track your jogs, a `Description` to describe the jog and `CreatedAt` property for an audit trail.

> Next, create an ApplicationDbContext class inside of a new folder called Data. For now, it will be pretty straightforward:

Reviewer: ```suggestion
Next, create an `ApplicationDbContext` class inside of a new folder called `Data`. For now, it will be pretty straightforward:
```

  //  to avoid creating duplicate seed data.
}
```

## Add Authentication to Your Angular Application

Handling the authentication is never an easy or comfortable task. If you want to quickly and easily take care of authentication for your application you're likely in the market for a solution like Okta. That's why so many developers love and use Okta, they never want to build auth again and are happy to leave it to us.

To get started with Okta, sign up for a [forever-free developer account](https://developer.okta.com/signup/) or (log in if you already have one).

{% img blog/mvc-angular/okta-signup.png alt:"okta signup" width:"800" %}{: .center-image }

You should see the Dashboard and in the upper right corner, there should be your unique Org URL. Save it for later.

{% img blog/mvc-angular/okta-org-url.png alt:"okta org url" width:"800" %}{: .center-image }

Now you need to create a new application by browsing to the Applications tab and clicking Add Application, and from the first page of the wizard choose **Single-Page App**.

On the settings page, enter `OktaMvcAngular` as your name value and select Implicit(Hybrid).

> {% img blog/mvc-angular/okta-app-settings.png alt:"okta application settings" width:"800" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/mvc-angular/okta-app-settings.png alt:"okta application settings" width:"700" %}{: .center-image }
```

Now that your application has been created copy down the Client ID and Client secret values on the following page, you'll need them soon (of course, yours will be different).

> {% img blog/mvc-angular/okta-client-credentials.png alt:"okta client credentials" width:"400" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/mvc-angular/okta-client-credentials.png alt:"okta client credentials" width:"500" %}{: .center-image }
```

```cs
using Newtonsoft.Json.Serialization;
```

You will also want to specify the port for your ASP.NET application. You need to know the exact address so you can communicate with the API from your Angular application. To do so, right click on the project in the solution explorer and click **properties**. In the main properties window, choose **Web** from the left-hand menu and set the **Project Url** property to `http://localhost:8080`.

## Set Up the Angular Application for ASP.NET MVC
Since you will let your ASP.NET MVC application handle routing you can remove the `app-routing.module.ts` file and its references in the `AppModule`. Your `app.module.ts` file should look like this:

```ts
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

After that is done, delete the `<router-outlet></router-outlet>` code from `app.component.html`.

> You also need to make sure that the Angular CLI outputs files to the appropriate folder. Inside of your Angular folder open `angular.json file and change the `outputPath` setting:

Reviewer: ```suggestion
You also need to make sure that the Angular CLI outputs files to the appropriate folder. Inside of your Angular folder open `angular.json` file and change the `outputPath` setting:
```

Your app contains only one root module, `AppModule`, and one component `AppComponent`. Since the application that you will be making here is a small one you will not create additional modules. Once the application grows and gets more complicated, you should introduce new modules, shared modules to organize your code better. The application itself will only show some records from the API. However, since your Angular application is interacting with the server, you should create a separate Angular service for this purpose. 

You can also create a TypeScript model for the `JoggingRecord`. First create a `shared` folder inside the `app` folder, which is part of the Angular application that gets created by the Angular CLI. 

Inside of the `shared` folder create a new file — `JoggingRecord.ts`:

```ts
export default class JoggingRecord {
  id: number;
  distance: number;
  description: string;
  createdAt: string;
}
```

The class above will allow you to have a strongly typed model for the `JoggingRecord` inside of your Angular application.

You can now create the API service inside of the `shared` folder. Name the file `api.service.ts`:

> {% raw %}

Reviewer: This is only needed if you have double braces, e.g., `{{ variable }}`

```ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import JoggingRecord from './JoggingRecord';

> @Injectable()

Reviewer: If you change this to `@Injectable({providedIn: 'root'})`, you don't need to register it as a provider in `app.module.ts`.

export default class ApiService {
  public API = 'http://localhost:8080/api';
  public JOGGING_RECORDS_ENDPOINT = `${this.API}/joggingrecords`;

  constructor(private http: HttpClient) { }

  getAll(): Observable<Array<JoggingRecord>> {
    return this.http.get<Array<JoggingRecord>>(this.JOGGING_RECORDS_ENDPOINT);
  }
}
```
{% endraw %}

The `ApiService` class is the bridge between our Angular application and our Web API endpoints. You need it to fetch the data from our server.

Add the service to the App module by updating the `providers` array inside of `app.module.ts` file. You also need to import the `HttpClientModule`. Your `app.module.ts` file should end up  looking like this:

```ts
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';

import { AppComponent } from './app.component';
import ApiService from '../shared/api.service';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    HttpClientModule,
    BrowserModule,
  ],
  providers: [ApiService],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

## Fetch Data from the ASP.NET API Endpoint

> Update the `app.component.ts` file with the following content

Reviewer: ```suggestion
Update the `app.component.ts` file with the following content:
```

```ts
import { Component, OnInit } from '@angular/core';

import JoggingRecord from '../shared/JoggingRecord';
import ApiService from '../shared/api.service';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit  {
  joggingRecords: Array<JoggingRecord>;

  constructor(private apiService: ApiService) {
  }

  ngOnInit() {
    this.apiService.getAll().subscribe(data => {
      this.joggingRecords = data;
    });
  }
}

```

The code above will subscribe will assign the data that `ApiService` fetches from server to the `joggingRecords` property. 

> You also need to update the template for this component by updating `app.component.html` file:

Reviewer: ```suggestion
You also need to update this component's template (`app.component.html`):
```

    <br />
  </span>
</div>
```
{% endraw %}

## Test Out Your ASP.NET and Angular Application
Since you already have the Angular CLI outputting the code changes for your Angular application in the `Bundles/AngularOutput` folder, and your MVC application is making use of that folder you can simply start the MVC application and preview Angular app inside of it.

From your Visual Studio use `CTRL+F5` to start the application your browser should open and you should see the following screen:

{% img blog/mvc-angular/app-homepage.png alt:"application home page" width:"800" %}{: .center-image }

Clicking on the Angular link in the menu will actually render the MVC view for the Angular application and from that point it's Angular who is in charge of rendering. The Angular page should look like this:

{% img blog/mvc-angular/app-jogging-records.png alt:"application jogging records" width:"800" %}{: .center-image }

If you open the Dev Tools on your editor you should see that the Angular application actually hits the ASP.NET API:

{% img blog/mvc-angular/app-dev-tools.png alt:"application dev tools" width:"800" %}{: .center-image }

## Learn More About ASP.NET Core and Angular

> You can find the source code for the complete application on GitHub: [https://github.com/oktadeveloper/aspnet-mvc-angular-example](https://github.com/oktadeveloper/aspnet-mvc-angular-example).

Reviewer: This repo needs a name and URL at the top.

If you enjoyed building this ASP.NET MVC with Angular, check out more full-stack application posts from Okta.

> *[Build a CRUD App with ASP.NET Framework 4.x Web API and Angular](https://developer.okta.com/blog/2018/07/27/build-crud-app-in-aspnet-framework-webapi-and-angular)

Reviewer: ```suggestion
* [Build a CRUD App with ASP.NET Framework 4.x Web API and Angular](/blog/2018/07/27/build-crud-app-in-aspnet-framework-webapi-and-angular)
```

> * [Angular 6 – What's New and Why Upgrade?](https://developer.okta.com/blog/2018/05/09/upgrade-to-angular-6)

Reviewer: ```suggestion
* [Angular 7: What's New and Noteworthy + OIDC Goodness](/blog/2018/12/04/angular-7-oidc-oauth2-pkce)
```

> * [Token Authentication in ASP.NET Core – A Complete Guide](https://developer.okta.com/blog/2018/03/23/token-authentication-aspnetcore-complete-guide)

Reviewer: ```suggestion
* [Token Authentication in ASP.NET Core – A Complete Guide](/blog/2018/03/23/token-authentication-aspnetcore-complete-guide)
```

> * [Build a Secure CRUD App with ASP.NET Core and React](https://developer.okta.com/blog/2018/07/02/build-a-secure-crud-app-with-aspnetcore-and-react)

Reviewer: ```suggestion
* [Build a Secure CRUD App with ASP.NET Core and React](/blog/2018/07/02/build-a-secure-crud-app-with-aspnetcore-and-react)
```
